HiP-UserStore/Controllers/UsersController.cs

```diff
@@ -0,0 +1,267 @@
+                .AsQueryable()
+                .FirstOrDefault(u => u.Email == email);
+
+            if (user == null)
+                return NotFound();
+
+            var result = new UserResult(user)
+            {
+                Roles = await Auth.GetUserRolesAsStringAsync(user.UserId, _authConfig),
+                ProfilePicture = GenerateFileUrl(user.UserId)
+            };
+
+            return Ok(result);
+        }
+
+        [HttpGet("Me")]
+        [ProducesResponseType(typeof(UserResult), 200)]
+        [ProducesResponseType(400)]
+        [ProducesResponseType(404)]
+        public async Task<IActionResult> GetSelfAsync() => await GetByIdAsync(User.Identity.GetUserIdentity());
+
+        [HttpPut("{userId}")]
+        [ProducesResponseType(204)]
+        [ProducesResponseType(400)]
+        [ProducesResponseType(403)]
+        [ProducesResponseType(404)]
+        public async Task<IActionResult> PutAsync(string userId, [FromBody]UserUpdateArgs args)
+        {
+            if (!ModelState.IsValid)
+                return BadRequest(ModelState);
+
+            if (!_userIndex.TryGetInternalId(userId, out var internalId))
+                return NotFound();
+
+            if (!UserPermissions.IsAllowedToModify(User.Identity, userId))
+                return Forbid();
+
+            var ev = new UserUpdated
+            {
+                Id = internalId,
+                UserId = User.Identity.GetUserIdentity(), // Note: refers to the API caller, not the updated user
+                Properties = args,
+                Timestamp = DateTimeOffset.Now
+            };
+
+            await _eventStore.AppendEventAsync(ev);
+            return NoContent();
+        }
+
+        [HttpPost("Invite")]
+        public IActionResult InviteUsers(/*InviteArgs args*/)
+        {
+            throw new NotImplementedException(); // TODO: Migrate from CmsWebApi
+        }
+
+        /// <summary>
+        /// This method is intended to be called by Auth0.
+        /// A hook should be configured in Auth0 that calls this API after a user signed up.
+        /// </summary>
+        [HttpPost]
+        [ProducesResponseType(400)]
+        [ProducesResponseType(409)] // User with same ID already exists
+        [ProducesResponseType(201)]
+        public async Task<IActionResult> RegisterAsync([FromBody]UserArgsAuth0 args)
+        {
+            if (!ModelState.IsValid)
+                return BadRequest(ModelState);
+
+            // Note: args.Id contains the ID of the new user, User.Identity.GetUserIdentity() is something else
+            // (because the HTTP request is not initiated by the user, but by the Auth0 system)
+            var userId = "auth0|" + args.Id;
+
+            if (_userIndex.TryGetInternalId(userId, out var _))
+                return StatusCode(409, new { Message = $"A user with ID '{userId}' already exists" });
+
+            var ev = new UserCreated
```

Comment:
Why we are not storing Username from UserArgsAuth0 ?

Reply:
At the moment there is no way to enter a username during sign-up with Auth0, so the username is always just the email address. Furthermore: "User" doesn't have a username field, only FirstName & LastName... and I don't think either of those should be assigned an email address.